spring-security-oauth2/src/main/java/org/springframework/security/oauth2/provider/code/RedisAuthorizationCodeServices.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2006-2016 the original author or authors.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ * 
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.springframework.security.oauth2.provider.code;
+
+import java.util.List;
+
+import org.springframework.data.redis.connection.RedisConnection;
+import org.springframework.data.redis.connection.RedisConnectionFactory;
+import org.springframework.security.oauth2.provider.OAuth2Authentication;
+import org.springframework.security.oauth2.provider.code.RandomValueAuthorizationCodeServices;
+import org.springframework.security.oauth2.provider.token.store.redis.JdkSerializationStrategy;
+import org.springframework.security.oauth2.provider.token.store.redis.RedisTokenStoreSerializationStrategy;
+
+/**
+ * Implementation of authorization code services that stores the codes and authentication in Redis.
+ * 
+ * @author Stefan Rempfer
+ */
+public class RedisAuthorizationCodeServices extends RandomValueAuthorizationCodeServices {
+
+	private static final String AUTH_CODE = "auth_code:";
+
+	private final RedisConnectionFactory connectionFactory;
+
+	private String prefix = "";
+
+	private RedisTokenStoreSerializationStrategy serializationStrategy = new JdkSerializationStrategy();
+
+	/**
+	 * Default constructor.
+	 *
+	 * @param connectionFactory the connection factory which should be used to obtain a connection to Redis
+	 */
+	public RedisAuthorizationCodeServices(RedisConnectionFactory connectionFactory) {
+		this.connectionFactory = connectionFactory;
+	}
+
+	@Override
+	protected void store(String code, OAuth2Authentication authentication) {
+		byte[] key = serializeKey(AUTH_CODE + code);
+		byte[] auth = serialize(authentication);
+
+		RedisConnection conn = getConnection();
+		try {
+			conn.set(key, auth);
```

Comment:
By default the code is a randomly generated 6-letter string. It's possible to generate two identical codes, especially in a system that handles a high degree of volume. 

This implementation of the `store` method will quietly overwrite the existing code stored in Redis. I believe a better approach is to use SETNX command instead (https://redis.io/commands/setnx) and verify that the code is actually persisted without collision (you get a `Boolean` from setNX call see https://docs.spring.io/spring-data/redis/docs/current/api/org/springframework/data/redis/connection/RedisStringCommands.html#setNX-byte:A-byte:A-).